src/aapns/connection.py

```diff
@@ -0,0 +1,440 @@
+    # * closing and not closed: graceful shutdown
+    # * closed: totally dead
+    # * fixme: __aexit__() finished
+    closed = closing = False
+    outcome: str = None
+    r = w = bgr = bgw = None
+    channels: Dict[int, Channel]
+    max_concurrent_streams = 100  # recommended in RFC7540#section-6.5.2
+    last_new_sid = last_sent_sid = -1  # client streams are odd
+
+    def __repr__(self):
+        bits = [self.state, f"{self.host}:{self.port}"]
+        if self.state != "closed":
+            bits.extend([f"buffered:{self.buffered}", f"inflight:{self.inflight}"])
+        else:
+            bits.append(f"outcome:{self.outcome}")
+        return "<Connection %s>" % " ".join(bits)
+
+    def __init__(self, origin: str, ssl=None):
+        self.channels = dict()
+        url = urlparse(origin)
+        assert url.scheme == "https"
+        assert url.hostname
+        assert not url.username
+        assert not url.password
+        assert not url.path
+        assert not url.params
+        assert not url.query
+        assert not url.fragment
+        self.host = url.hostname
+        self.port = url.port or 443
+        self.ssl = ssl if ssl else create_ssl_context()
+
+    async def __aenter__(self):
+        assert ssl.OP_NO_TLSv1 in self.ssl.options
+        assert ssl.OP_NO_TLSv1_1 in self.ssl.options
+        # https://bugs.python.org/issue40111 validate h2 alpn
+
+        self.please_write = Event()
+        self.r, self.w = await open_connection(
+            self.host, self.port, ssl=self.ssl, ssl_handshake_timeout=5
+        )
+        info = self.w.get_extra_info("ssl_object")
+        assert info, "HTTP/2 server is required"
+        proto = info.selected_alpn_protocol()
+        assert proto == "h2", "Failed to negotiate HTTP/2"
+
+        self.conn = h2.connection.H2Connection(
+            h2.config.H2Configuration(client_side=True, header_encoding="utf-8")
+        )
+
+        self.conn.local_settings = h2.settings.Settings(
+            client=True,
+            initial_values={
+                h2.settings.SettingCodes.ENABLE_PUSH: 0,
+                # FIXME Apple server settings:
+                # HEADER_TABLE_SIZE 4096
+                # MAX_CONCURRENT_STREAMS 1000
+                # INITIAL_WINDOW_SIZE 65535
+                # MAX_FRAME_SIZE 16384
+                # MAX_HEADER_LIST_SIZE 8000
+                h2.settings.SettingCodes.MAX_CONCURRENT_STREAMS: 2 ** 20,
+                h2.settings.SettingCodes.MAX_HEADER_LIST_SIZE: 2 ** 16 - 1,
+                h2.settings.SettingCodes.INITIAL_WINDOW_SIZE: MAX_RESPONSE_SIZE,
+            },
+        )
+
+        self.conn.initiate_connection()
+        # APN response body is empty (ok) or small (error)
+        self.conn.increment_flow_control_window(CONNECTION_WINDOW_SIZE)
+        self.please_write.set()
+
+        self.bgr = create_task(self.background_read(), name="bg-read")
+        self.bgw = create_task(self.background_write(), name="bg-write")
```

Comment:
why are these background tasks needed?

Reply:
Because HTTP/2 is a complex protocol 😉

Ultimately, there are 2 ways to deadlock:
* remote (trying to write, blocking because buffer is full, when a read is needed), or
* local (trying to read, blocking as there's no data, when a write is needed)

Thing is, `h2` will automatically respond to some data received from the server, thus a write may be required after each read. 

Likewise, the server may automatically respond to some data we send, thus a read may be required after each write.

So, to make this simpler, I've decided to have 2 separate tasks, for read and write.

Additionally, the connection may be in either of the half-closed states... 😈

---

Obv, there are other ways to solve this. I just picked this way.

Reply:
> Thing is, `h2` will automatically respond to some data received from the server, thus a write may be required after each read.

I've solved this by writing after read (if needed). basically after calling something on `H2Connection` check if there is data to send and send it. at least `background_write` is probably not needed even in your approach since `data_to_send` will only change if you call something on `H2Connection`. 

My concern with background tasks is that they make error handling/detection so much harder and in an extreme case a broken background task can turn a connection into a dead state without the connection knowing about it. I guess for reading the background task is still needed though especially if there's a time of few/no requests on a connection and the server is sending control frames.

Reply:
That's how I started too, but then writing may block, so it needs `await` and during that time the state can change, furthermore, concurrent requests may try to write at the same time time.
`httpx` uses a mutex to address that.
But then, what happens if the writing request times out? What is it's cancelled?
Ok, sure, I'll shield it, but then how to pass the baton to the next writer?
(I believe on the issues you've opened against `httpx` is exactly due to this).
Maybe relying on the mutex could be enough, I dunno...

I've started with `httpx` way, but soon realised that it's just so error-prone, while the background read is certainly needed (because server changes settings asynchronously after connection is created, because connection may get closed, etc.)...

Thus, instead of spreading `if data := xxx(): await write(data)` all over the codebase, I've decided to concentrate all writes in one place.

I've started with single background task (read+write), but it was getting a bit too complex, so I've split them up.

Your concern is totally valid.

My concern is that I need to be able to understand my damned code :orz: